### ISSUE TITLE
Selecting family as one object

### ISSUE THREAD

**Issue by pop** [aevyrie/bevy_mod_picking#108]
I'd love to be able to group multiple Meshes into one Pickable entity. i.e., if I spawn a Parent entity with multiple children meshes, selecting any one of the children selects the entire hierarchy (parent and siblings).

My use-case is that I am importing a GLTF Scene which has Nodes which are made of multiple Meshes. I want to treat each GLTF Node as one Pickable entity. I've tried implementing this locally but haven't made progress.

If you (A) know of a workaround or (B) know where in the code I could add this feature I would gladly make this a pull request.

I expect this would be implemented with the following psuedocode:

1. When a mesh is selected, check if the mesh has a parent.
2. Update the state for all of that parent's children to match the selected child's state.
3. Hierarchy's could be implemented with a `PickableHierarchyBundle` on the parent and/or children -- or just tag the parent (as well as the children) with the `PickableBundle`.

I'd love to discuss design if I'm [not] on the right track.

P.S., I love the plugin! I'm using it in a game and it's made a point-and-click mechanic very easy!

**Comment by pop** [comment]
Update, I got a very rough and dirty version of this feature implemented. I'd love feedback on the idea and the implementation.

[Branch Here](https://github.com/aevyrie/bevy_mod_picking/compare/master...pop:pop/pickable-hierarchy)

**Comment by aevyrie** [comment]
Thanks @pop, sorry for the delayed response. This looks really useful! If you're still willing to implement this, I could help out on a PR branch.

**Comment by pop** [comment]
I just picked up my gamedev project again, I should have a PR up for this in a few days.

**Comment by aevyrie** [comment]
Cool! Looking forward to seeing what you come up with. 🙂

**Comment by sildude** [comment]
Would love to be able to simply spawn a gltf scene and have this work out of the box.

**Comment by ericvrp** [comment]
Is there any news about this?

**Comment by pop** [comment]
@ericvrp I struggled with the implementation details but I finally got a _rough_ implementation done in #126. Check it out if you're interested, I'd love feedback on the API choices and implementation.

**Comment by pop** [comment]
@sildude The implementation in #126 doesn't work with selecting GLTFs, but after digging into the code I see how one _would_ add GLTF selections.

The interface would require tagging a parent entity with `PickableMesh` or a new `PickableGltf` and spawning your GLTF mesh/scene as a child of that entity. Something like this:

```rust
let scene = assets.load("whatever.gltf#Scene0");

// cube
commands.spawn_bundle((
    Transform::from_xyz(0.0, 0.0, 0.0),
    GlobalTransform::identity(),
))
.with_children(|parent| {
    parent.spawn_scene(scene);
})
.insert_bundle(PickableGltf::default()); // <- Makes the scene one pickable object
```

Once #126 is finalized I'll start on adding GLTF picking. I agree it'd be a useful addition.

**Comment by ericvrp** [comment]
I know you are not finished and probably I am doing something silly anyway (replacing PickableGltf with PickableMesh in your example above) but I get this error:

<img width="1243" alt="image" src="https://user-images.githubusercontent.com/4216292/163830245-65339eb2-d19d-452e-b6f6-ee31cc84df78.png">

**Comment by pop** [comment]
@ericvrp on my PR branch check out the [`family.rs` example](https://github.com/aevyrie/bevy_mod_picking/pull/126/files#diff-868a5bb33337d2003ff530bb03478fc32cbb929fc551f475dc791238ad640031). Like I said, GLTF selection doesn't work but that example shows you what is supported by the current implementation.

**Comment by ericvrp** [comment]
Is this PickableGltf something we can hope for in an upcoming version?

**Comment by aevyrie** [comment]
This is working in the `refactor` branch with event bubbling. It's much more flexible and useful than a family-specific solution.

**Comment by aevyrie** [comment]
So, the concept that the `refactor` branch adds is event bubbling. This event propagation technique allows events to "bubble" up a hierarchy and trigger events on event listeners.

For the use case outlined in this ticket, you would:

1. make every mesh in your scene pickable
2. add an event forwarder (event listener) on each of the gltf nodes you want to control selection, watching for the PointerClick event
3. When one of the forwarded events is received, go through all children of the listener and select them

**Comment by KnstAnt** [comment]
I have added to my project what you are discussing. But I implemented it via gltf_scene_add_event and I don't see anything similar in the crate where I could add my code. Otherwise, there is no difficulty, first go through the children and add a PickableBundle component for everyone who has a mesh. Then, when triggered, the SelectionEvent will climb up the parents to the root and set the required Interaction to it.
